### PR TITLE
More interpreter optimizations

### LINKED ISSUE #69: Make API methods names clearer : comply to abstract-encoding
https://github.com/mafintosh/abstract-encoding is nice and it looks like protodef createPacketBuffer and parsePacketBuffer could fit that api.

### PR DESCRIPTION
**Pull request by Saiv46**
Closes #69 and #75

Benchmark results by @Karang (benchmark_all_types):
```markdown
**ProtoDef-io:master**:
read x 26,817 ops/sec ±1.43% (92 runs sampled)
  ✓ reads (5617ms)
write x 16,188 ops/sec ±0.46% (96 runs sampled)
  ✓ writes (5788ms)
read (compiled) x 104,248 ops/sec ±1.56% (87 runs sampled)
  ✓ reads (compiled) (5474ms)
write (compiled) x 79,582 ops/sec ±0.39% (95 runs sampled)
  ✓ writes (compiled) (5616ms)
**Saiv46:master**:
read x 67,847 ops/sec ±0.45% (95 runs sampled)
  ✓ reads (5676ms)
write x 40,270 ops/sec ±0.38% (95 runs sampled)
  ✓ writes (5558ms)
read (compiled) x 102,333 ops/sec ±1.02% (93 runs sampled)
  ✓ reads (compiled) (5524ms)
write (compiled) x 77,432 ops/sec ±1.04% (86 runs sampled)
  ✓ writes (compiled) (5614ms)
```

### CODE REVIEW

File: src/interpreter.js [hunk 1]
```diff
@@ -1,58 +1,17 @@
...
-  }
-  return [function read (buffer, offset, typeArgs, context) {
-    return functions[0].call(this, buffer, offset, produceArgs(typeArgs), context)
-  }, function write (value, buffer, offset, typeArgs, context) {
-    return functions[1].call(this, value, buffer, offset, produceArgs(typeArgs), context)
-  }, function sizeOf (value, typeArgs, context) {
-    if (typeof functions[2] === 'function') { return functions[2].call(this, value, produceArgs(typeArgs), context) } else { return functions[2] }
-  }]
-}
+// FIXME: Increases performance but introduces unexpected bugs in future
```
Comment (Karang) [inline src/interpreter.js:7]: Can you elaborate on this comment ? What kind of unexpected bugs should we worry about ?

Comment (Saiv46) [reply src/interpreter.js:7]: https://github.com/ProtoDef-io/node-protodef/pull/107/files/43f7401c0be36b6d001db2c18a24ad8d7c91b34a#diff-14cffeee7a34687f7a373b01152975d8R135
Before this PR the interpreter every time does copy of datatype and merges it with passed arguments, I don't know why this behavior was there. Maybe @rom1504 should explain that to me.

**Comment by rom1504** [comment]
Can you open a PR for node-minecraft-protocol using the interpreter and this PR as protodef version ?
This would make it possible to check better that this works

**Comment by rom1504** [comment]
Could you explain what you did here and why it is faster ?

File: src/interpreter.js (hunk 1 above)
Comment (Saiv46) [reply src/interpreter.js]: That kind of bugs:
https://github.com/PrismarineJS/node-minecraft-protocol/pull/731/checks?check_run_id=777137891#step:6:239

**Comment by rom1504** [comment]
@Saiv46 do you intend to finish this?

**Comment by Saiv46** [comment]
I need to figure out what's missing here. Something sropped me from finishing this PR... Oh yes, fields need to be optimized...

**Comment by Saiv46** [comment]
> Could you explain what you did here and why it is faster ?

@rom1504 First I was just refactoring existing code - first by just splitting it, then applying some optimizations from `protodef-neo`.

Now I read [this article](https://www.reaktor.com/blog/javascript-performance-fundamentals-make-bluebird-fast/) about optimization fundamentals - now I'm doing it.

By the way

```diff
-  396 passing (1s)
+  396 passing (464ms)

- C:\Users\Admin\node-protodef>npx mocha ./benchmark/benchmark_all_types.js
+ D:\WORKING\Saiv46\node-protodef>npx mocha ./benchmark/benchmark_all_types.js

- read x 8,250 ops/sec ±2.34% (84 runs sampled)
-   √ reads (6102ms)
+ read x 20,366 ops/sec ±9.83% (78 runs sampled)
+   √ reads (5832ms)
- write x 4,229 ops/sec ±12.03% (76 runs sampled)
-   √ writes (5533ms)
+ write x 11,746 ops/sec ±22.16% (84 runs sampled)
+   √ writes (5698ms)
- read (compiled) x 46,755 ops/sec ±3.83% (85 runs sampled)
-   √ reads (compiled) (5757ms)
+ read (compiled) x 45,324 ops/sec ±5.09% (83 runs sampled)
+   √ reads (compiled) (5625ms)
- write (compiled) x 29,127 ops/sec ±3.31% (80 runs sampled)
-   √ writes (compiled) (5516ms)
+ write (compiled) x 29,561 ops/sec ±3.43% (83 runs sampled)
+   √ writes (compiled) (5632ms)

  4 passing (23s)
```

File: src/interpreter.js (hunk 1 above)
Comment (Saiv46) [reply src/interpreter.js:8]: Switched from `JSON.stringify` to `_.cloneDeep`, it was fast enough. Also fixed `defaultTypeArgs !== 'object'` to `typeof defaultTypeArgs !== 'object'` (how did I miss that before?)

**Comment by Saiv46** [comment]
>3x interpreter performance improvement achieved!

```
read x 23,910 ops/sec ±6.52% (80 runs sampled)
  √ reads (6062ms)
write x 12,641 ops/sec ±6.70% (81 runs sampled)
  √ writes (5764ms)
```

Now I should undo compiler changes (because this is out of the scope of this PR)

**Comment by Saiv46** [comment]
I was tried to make something more on this PR, but I run out of ideas that will not break anything.

**Pull request closed without merging**